internal/envoy/versions.go

```diff
@@ -20,30 +20,122 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"strconv"
+	"strings"
 
+	"github.com/tetratelabs/func-e/internal/globals"
 	"github.com/tetratelabs/func-e/internal/version"
 )
 
-// FuncEVersions returns a version map from a remote URL. eg globals.DefaultEnvoyVersionsURL.
-func FuncEVersions(ctx context.Context, envoyVersionsURL string, p version.Platform, v version.Version) (version.ReleaseVersions, error) {
+type funcEVersionsGetter func() (version.ReleaseVersions, error)
+
+// FuncEVersions fetches a version map form a remote URL and gets the latest patched version given
+// a version input. This input can be major.minor.patch (e.g. 1.1.1) or major.minor (1.1).
+type FuncEVersions interface {
+	// Get returns a version map from a remote URL. e.g. from globals.DefaultEnvoyVersionsURL.
+	Get() (version.ReleaseVersions, error)
+
+	// FindLatestVersions returns the latest patch version given an Envoy version.
+	FindLatestVersions(version.Version) (string, error)
+}
+
+// NewFuncEVersions creates a new FuncEVersions to fetch a version map from a remote URL.
+func NewFuncEVersions(ctx context.Context, envoyVersionsURL string, p version.Platform, v version.Version) FuncEVersions {
+	return &funcEVersions{ctx: ctx, envoyVersionsURL: envoyVersionsURL, platform: p, version: v}
+}
+
+type funcEVersions struct {
+	ctx              context.Context
+	envoyVersionsURL string
+	platform         version.Platform
+	version          version.Version
+
+	// Allow to override Get implementation.
+	getFunc funcEVersionsGetter
+}
+
+// Get returns a version map from a remote URL. e.g. from globals.DefaultEnvoyVersionsURL.
+func (f funcEVersions) Get() (version.ReleaseVersions, error) {
 	result := version.ReleaseVersions{}
 	// #nosec => This is by design, users can call out to wherever they like!
-	resp, err := httpGet(ctx, envoyVersionsURL, p, v)
+	resp, err := httpGet(f.ctx, f.envoyVersionsURL, f.platform, f.version)
 	if err != nil {
 		return result, err
 	}
 	defer resp.Body.Close() //nolint
 
 	if resp.StatusCode != http.StatusOK {
-		return result, fmt.Errorf("received %v status code from %v", resp.StatusCode, envoyVersionsURL)
+		return result, fmt.Errorf("received %v status code from %v", resp.StatusCode, f.envoyVersionsURL)
 	}
 	body, err := io.ReadAll(resp.Body) // fully read the response
 	if err != nil {
-		return result, fmt.Errorf("error reading %s: %w", envoyVersionsURL, err)
+		return result, fmt.Errorf("error reading %s: %w", f.envoyVersionsURL, err)
 	}
 
 	if err := json.Unmarshal(body, &result); err != nil {
 		return result, fmt.Errorf("error unmarshalling Envoy versions: %w", err)
 	}
+
 	return result, nil
 }
+
+// FindLatestVersions returns the latest patch version given an Envoy version.
+func (f funcEVersions) FindLatestVersions(envoyVersion version.Version) (string, error) {
+	if matched := globals.EnvoyVersionPattern.MatchString(string(envoyVersion)); matched {
```

Comment:
ex we shouldn't enter this case at all, there's no reason to assume our internal code will pass the wrong value, and if we did we'd have to test it, so it isn't really fun

This type shouldn't even need to be instantiated if someone passes a concrete version iotw.

Reply:
Got it.